pkg/inspect/modules_test.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inspect_test
+
+import (
+	"hpc-toolkit/pkg/inspect"
+	"hpc-toolkit/pkg/modulereader"
+	"log"
+	"path/filepath"
+	"testing"
+
+	"golang.org/x/exp/slices"
+)
+
+type varInfo = modulereader.VarInfo
+type predicate = func(modInfo) bool
+
+type modInfo struct {
+	modulereader.ModuleInfo
+	inspect.SourceAndKind
+}
+
+func (m *modInfo) Input(name string) (varInfo, bool) {
+	ind := slices.IndexFunc(m.Inputs, func(v varInfo) bool { return v.Name == name })
+	if ind == -1 {
+		return varInfo{}, false
+	}
+	return m.Inputs[ind], true
+}
+
+var allMods []modInfo = nil
+
+func getModules() []modInfo {
+	if allMods != nil {
+		return allMods
+	}
+	sks, err := inspect.LocalModules()
+	if err != nil {
+		log.Fatal(err)
+	}
+	allMods = []modInfo{}
+	for _, sk := range sks {
+		info, err := modulereader.GetModuleInfo(filepath.Join("../..", sk.Source), sk.Kind)
+		if err != nil {
+			log.Fatal(err)
+		}
+		allMods = append(allMods, modInfo{ModuleInfo: info, SourceAndKind: sk})
+	}
+	return allMods
+}
+
+func query(p predicate) []modInfo {
+	ret := []modInfo{}
+	for _, mod := range getModules() {
+		if p(mod) {
+			ret = append(ret, mod)
+		}
+	}
+	return ret
+}
+
+func all(ps ...predicate) predicate {
```

Comment:
`all`  is super-predicate that is conjunction of predicates-operands. Just like `and`, but `and` has notion of being binary operation (operation with two operands), while `all` deals with set of operands.

```
"Select large red apples" =>  
(size=large) and (color=red) and (type=apple) => 
all(size("large"), color("red"), type("apple"))
```

See: https://docs.python.org/3/library/functions.html#all

The result of `all` with no operands is just `true`, because:
```
all(head, tail...) := head and all(tail)
all(head) := head and all()
all() := true
```

This lets as have a nice `query(all())` that literally means "query all"